drivers: spi: add spi-bitbang driver

marc-cpdesign opened the pull request:
This adds an spi master mode driver via bitbanged gpio. The driver
supports the various clock polarity and phase configurations, and can
also work with word sizes which are non multiples of 8bits, currently up
to 16 bits.

A sample program is also added demonstrating basic use of the driver with
9bit data words.

Signed-off-by: Marc Reilly <marc@cpdesign.com.au>

tbursztyka reviewed (approved):
nice

mbolivar-nordic commented on samples/drivers/spi_bitbang/boards/nrf52840dk_nrf52840.overlay:
```diff
@@ -0,0 +1,20 @@
+		label = "SPIBB";
+		compatible = "zephyr,spi-bitbang";
+		status="okay";
+		reg=<0 0x1000>;
```
No need for `reg` or a unit address for these nodes, I think. You should be able to define these nodes without either if they're children of the root node as is being done here. See `/pwmleds` for example.

marc-cpdesign replied:
I'm admittedly a bit of a fumbler for setting up device tree stuff, earlier code wouldn't work without the "reg" there and it just stayed :|   Fixed now.

mbolivar-nordic commented on samples/drivers/spi_bitbang/README.rst:
```diff
@@ -0,0 +1,54 @@
+********************
+
+The application will build only for a target that has a :ref:`devicetree
+<dt-guide>` entry with ``zephyr,spi-bitbang`` as a compatible.
```
```suggestion
<dt-guide>` entry with :dtcompatible:`zephyr,spi-bitbang` as a compatible.
```

This is preferable because it creates a link to the binding's autogenerated documentation in the [bindings index](https://docs.zephyrproject.org/latest/reference/devicetree/bindings.html).

mbolivar-nordic commented on drivers/spi/spi_bitbang.c:
```diff
@@ -0,0 +1,293 @@
+		return -EIO;
+	}
+
+	if (!device_is_ready(config->mosi_gpio.port)) {
```
This needs a null check.

mbolivar-nordic commented on drivers/spi/spi_bitbang.c:
```diff
@@ -0,0 +1,293 @@
+		LOG_ERR("GPIO port for mosi pin is not ready");
+		return -ENODEV;
+	}
+	if (gpio_pin_configure_dt(&config->mosi_gpio, GPIO_OUTPUT_INACTIVE)) {
```
Please propagate original errors to the callers instead of swallowing and returning -EIO, here and elsewhere.

mbolivar-nordic commented on drivers/spi/spi_bitbang.c:
```diff
@@ -0,0 +1,293 @@
+
+#define SPI_BITBANG_INIT(inst)						\
+	static struct spi_bitbang_config spi_bitbang_config_##inst = {	\
+		.clk_gpio = GPIO_DT_SPEC_INST_GET(0, clk_gpios),	\
+		.mosi_gpio = GPIO_DT_SPEC_INST_GET_OR(0, mosi_gpios, {0}),	\
+		.miso_gpio = GPIO_DT_SPEC_INST_GET_OR(0, miso_gpios, {0}),	\
```
All of the first `0` arguments to each of these should be `inst`.

mbolivar-nordic reviewed (changes requested):
Thanks! I found a few driver model and DT things on a first pass review.

mbolivar-nordic reviewed (approved):
Thanks for your patience. The DT usage in the driver looks great now.

tbursztyka commented:
Can you rebase against main branch and push force again?

marc-cpdesign commented:
I've rebased on current master and pushed.. There were some running errors with the builder "./scripts/twister: not found"; it also still says that there are conflicts that need to be resolved but these were fixed in the rebase.
Please let me know if there's anything more I should do.

XenuIsWatching commented on drivers/spi/spi_bitbang.c:
```diff
@@ -0,0 +1,301 @@
+		}
+
+		while (shift >= 0) {
+			const int d = (w >> shift) & 0x1;
```
Since this is only an MSB first transfer... shouldn't it be checking that it's not being configured for SPI_TRANSFER_LSB.

XenuIsWatching commented on drivers/spi/spi_bitbang.c:
```diff
@@ -0,0 +1,301 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_spi_bitbang
```
Is this the intent of the zephyr vendor prefix? There is another very similar driver with the name gpio_i2c. Should this be named gpio_spi to be consistent?

https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2c/i2c_gpio.c

marc-cpdesign replied:
It made sense at the time :) but TBH I'm not sure.

There is also the "zephyr,2ic-emul" which has `#define DT_DRV_COMPAT zephyr_i2c_emul_controller` ... having a brief look through the binding and driver files, I get the vibe that the newer drivers use 'zephyr' as the vendor prefix, and the gpio_i2c and gpio_led etc do it a different way because they are older than the existing convention but remain the same for compatibility.

I don't really care though .. if there is a consensus that some other way is better I can change it.

mbolivar-nordic replied:
I think `zephyr` is an OK vendor prefix here following the rules laid out in https://docs.zephyrproject.org/latest/guides/dts/bindings.html#vendor-prefixes. But maybe @galak disagrees.

marc-cpdesign committed:
drivers: spi: add spi-bitbang driver

This adds an spi master mode driver via bitbanged gpio. Only syncronous
transfers are implemented. Clock signal timing is accomplished via busy
waits, the gpios are manipulated via the standard gpio interface; these
two factors limit the frequency at which it can operate - but here
a simple and generic implementation was chosen over performance.

The driver supports the various clock polarity and phase
configurations, and can also work with word sizes which are non
multiples of 8bits, currently up to 16 bits.

A sample program is also added demonstrating basic use of the driver
with 9bit data words.

Signed-off-by: Marc Reilly <marc@cpdesign.com.au>
